> 19-12-25 最后一次修改      
## 读取+清洗nan

In [1]:
import numpy as np
import netCDF4 as nc
import pandas as pd
import numpy.ma as ma

In [2]:
import matplotlib as mpl
import matplotlib.pyplot as plt

In [3]:
import xarray as xar

C:\ProgramData\Anaconda3\lib\site-packages\dask\config.py:168: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  data = yaml.load(f.read()) or {}
C:\ProgramData\Anaconda3\lib\site-packages\distributed\config.py:20: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  defaults = yaml.load(f)


In [4]:
file_path='./data/sanjioil.nc'

In [5]:
ds=nc.Dataset(file_path)

In [6]:
ds_xr=xar.open_dataset(file_path)

In [7]:
ds_xr.dims

Frozen(SortedKeysDict({'trajectory': 2958, 'time': 73}))

In [8]:
ds_xr.data_vars

Data variables:
    age_seconds                               (trajectory, time) float32 ...
    interfacial_area                          (trajectory, time) float32 ...
    mass_evaporated                           (trajectory, time) float32 ...
    water_fraction                            (trajectory, time) float32 ...
    y_wind                                    (trajectory, time) float32 ...
    sea_ice_area_fraction                     (trajectory, time) float32 ...
    density                                   (trajectory, time) float32 ...
    oil_film_thickness                        (trajectory, time) float32 ...
    x_sea_water_velocity                      (trajectory, time) float32 ...
    bulltime                                  (trajectory, time) float32 ...
    x_wind                                    (trajectory, time) float32 ...
    age_emulsion_seconds                      (trajectory, time) float32 ...
    sea_surface_wave_stokes_drift_y_velocity  (trajectory, t

In [9]:
ds_xr.coords

Coordinates:
  * trajectory  (trajectory) int32 1 2 3 4 5 6 ... 2953 2954 2955 2956 2957 2958
  * time        (time) datetime64[ns] 2018-01-14T22:20:00 ... 2018-01-17T22:20:00
    lon         (trajectory, time) float32 ...
    lat         (trajectory, time) float32 ...

获取所有的维度  
* 注意获取维度时，不可以使用下标进行索引，会报错

In [15]:
for temp in ds_xr.dims:
    print(type(temp))
    print(temp)

<class 'str'>
time
<class 'str'>
trajectory


In [10]:
for temp in ds_xr.coords:
    print(type(temp))
    print(temp)

<class 'str'>
trajectory
<class 'str'>
time
<class 'str'>
lon
<class 'str'>
lat


In [12]:
type(ds_xr.coords)

xarray.core.coordinates.DatasetCoordinates

使用.get判断是否存在指定的坐标名称

In [14]:
ds_xr.coords.get('xx')

In [11]:
ds_xr.coords['time']

<xarray.DataArray 'time' (time: 73)>
array(['2018-01-14T22:20:00.000000000', '2018-01-14T23:20:00.000000000',
       '2018-01-15T00:20:00.000000000', '2018-01-15T01:20:00.000000000',
       '2018-01-15T02:20:00.000000000', '2018-01-15T03:20:00.000000000',
       '2018-01-15T04:20:00.000000000', '2018-01-15T05:20:00.000000000',
       '2018-01-15T06:20:00.000000000', '2018-01-15T07:20:00.000000000',
       '2018-01-15T08:20:00.000000000', '2018-01-15T09:20:00.000000000',
       '2018-01-15T10:20:00.000000000', '2018-01-15T11:20:00.000000000',
       '2018-01-15T12:20:00.000000000', '2018-01-15T13:20:00.000000000',
       '2018-01-15T14:20:00.000000000', '2018-01-15T15:20:00.000000000',
       '2018-01-15T16:20:00.000000000', '2018-01-15T17:20:00.000000000',
       '2018-01-15T18:20:00.000000000', '2018-01-15T19:20:00.000000000',
       '2018-01-15T20:20:00.000000000', '2018-01-15T21:20:00.000000000',
       '2018-01-15T22:20:00.000000000', '2018-01-15T23:20:00.000000000',
       '2018-01-16T00:20:00.000000000', '2018-01-16T01:20:00.000000000',
       '2018-01-16T02:20:00.000000000', '2018-01-16T03:20:00.000000000',
       '2018-01-16T04:20:00.000000000', '2018-01-16T05:20:00.000000000',
       '2018-01-16T06:20:00.000000000', '2018-01-16T07:20:00.000000000',
       '2018-01-16T08:20:00.000000000', '2018-01-16T09:20:00.000000000',
       '2018-01-16T10:20:00.000000000', '2018-01-16T11:20:00.000000000',
       '2018-01-16T12:20:00.000000000', '2018-01-16T13:20:00.000000000',
       '2018-01-16T14:20:00.000000000', '2018-01-16T15:20:00.000000000',
       '2018-01-16T16:20:00.000000000', '2018-01-16T17:20:00.000000000',
       '2018-01-16T18:20:00.000000000', '2018-01-16T19:20:00.000000000',
       '2018-01-16T20:20:00.000000000', '2018-01-16T21:20:00.000000000',
       '2018-01-16T22:20:00.000000000', '2018-01-16T23:20:00.000000000',
       '2018-01-17T00:20:00.000000000', '2018-01-17T01:20:00.000000000',
       '2018-01-17T02:20:00.000000000', '2018-01-17T03:20:00.000000000',
       '2018-01-17T04:20:00.000000000', '2018-01-17T05:20:00.000000000',
       '2018-01-17T06:20:00.000000000', '2018-01-17T07:20:00.000000000',
       '2018-01-17T08:20:00.000000000', '2018-01-17T09:20:00.000000000',
       '2018-01-17T10:20:00.000000000', '2018-01-17T11:20:00.000000000',
       '2018-01-17T12:20:00.000000000', '2018-01-17T13:20:00.000000000',
       '2018-01-17T14:20:00.000000000', '2018-01-17T15:20:00.000000000',
       '2018-01-17T16:20:00.000000000', '2018-01-17T17:20:00.000000000',
       '2018-01-17T18:20:00.000000000', '2018-01-17T19:20:00.000000000',
       '2018-01-17T20:20:00.000000000', '2018-01-17T21:20:00.000000000',
       '2018-01-17T22:20:00.000000000'], dtype='datetime64[ns]')
Coordinates:
  * time     (time) datetime64[ns] 2018-01-14T22:20:00 ... 2018-01-17T22:20:00
Attributes:
    standard_name:  time
    long_name:      time

In [20]:
type(ds_xr.coords['time'].values)

numpy.ndarray

In [23]:
ds_xr.coords['time'].values[:5]

array(['2018-01-14T22:20:00.000000000', '2018-01-14T23:20:00.000000000',
       '2018-01-15T00:20:00.000000000', '2018-01-15T01:20:00.000000000',
       '2018-01-15T02:20:00.000000000'], dtype='datetime64[ns]')

In [25]:
ds_xr.coords['time'].values[:5].tolist()

[1515968400000000000,
 1515972000000000000,
 1515975600000000000,
 1515979200000000000,
 1515982800000000000]

## 获取`几个指定维度的`的`DataArray`  
如下可知：
    `mass_oil`有2958个值，该物理量有两个维度，分别为`trajectory`,`time`

In [11]:
# 读取一个 wind 
ds_xr.isel(time=1)['x_wind']

<xarray.DataArray 'x_wind' (trajectory: 2958)>
array([-6.335222e+00, -7.076474e+00, -5.017034e+00, ...,  9.969210e+36,
        9.969210e+36,  9.969210e+36], dtype=float32)
Coordinates:
  * trajectory  (trajectory) int32 1 2 3 4 5 6 ... 2953 2954 2955 2956 2957 2958
    time        datetime64[ns] 2018-01-14T23:20:00
    lon         (trajectory) float32 ...
    lat         (trajectory) float32 ...

In [10]:
ds_xr.isel(time=1)['mass_oil']

<xarray.DataArray 'mass_oil' (trajectory: 2958)>
array([1.624608e+03, 1.547655e+03, 1.677850e+03, ..., 9.969210e+36,
       9.969210e+36, 9.969210e+36], dtype=float32)
Coordinates:
  * trajectory  (trajectory) int32 1 2 3 4 5 6 ... 2953 2954 2955 2956 2957 2958
    time        datetime64[ns] 2018-01-14T23:20:00
    lon         (trajectory) float32 ...
    lat         (trajectory) float32 ...
Attributes:
    units:    kg

### 1- 尝试将xarray转为series

## 2 剔除掉nan的数据，只获取有效数据

#### 查看不同时刻的
并通过`where`剔除掉大于`max`以及`nan`的值  
使用xarray的dropna方法，注意与pandas中的方法有所区别

提示错误：
`ValueError: any must be a single dataset dimension`  
查看文档后发现`xarray.Dataset.dropna`尚不支持同时沿多个维度放置  

[参考](http://xarray.pydata.org/en/stable/generated/xarray.Dataset.dropna.html)  

考虑将xarray转换为dateframe，然后再清洗？  
[上网参考](https://stackoverflow.com/questions/52553925/python-xarray-remove-coordinates-with-all-missing-variables)
比较好的解决办法是：
1. 将`DataSet`->`DataFrame`
2. 在`DataFrame`中通过`pandas` `dropna` 即可

In [27]:
ds_xr.isel(time=30).get('x_wind')

<xarray.DataArray 'x_wind' (trajectory: 2958)>
array([ 9.969210e+36, -2.856885e+00,  9.969210e+36, ...,  9.969210e+36,
        9.969210e+36,  9.969210e+36], dtype=float32)
Coordinates:
  * trajectory  (trajectory) int32 1 2 3 4 5 6 ... 2953 2954 2955 2956 2957 2958
    time        datetime64[ns] 2018-01-16T04:20:00
    lon         (trajectory) float32 ...
    lat         (trajectory) float32 ...

In [40]:
xr_temp_x=ds_xr.isel(time=30)['x_wind']
xr_temp_x

<xarray.DataArray 'x_wind' (trajectory: 2958)>
array([ 9.969210e+36, -2.856885e+00,  9.969210e+36, ...,  9.969210e+36,
        9.969210e+36,  9.969210e+36], dtype=float32)
Coordinates:
  * trajectory  (trajectory) int32 1 2 3 4 5 6 ... 2953 2954 2955 2956 2957 2958
    time        datetime64[ns] 2018-01-16T04:20:00
    lon         (trajectory) float32 ...
    lat         (trajectory) float32 ...

In [41]:
xr_temp_y=ds_xr.isel(time=30)['y_wind']
xr_temp_y

<xarray.DataArray 'y_wind' (trajectory: 2958)>
array([9.969210e+36, 1.047276e+01, 9.969210e+36, ..., 9.969210e+36,
       9.969210e+36, 9.969210e+36], dtype=float32)
Coordinates:
  * trajectory  (trajectory) int32 1 2 3 4 5 6 ... 2953 2954 2955 2956 2957 2958
    time        datetime64[ns] 2018-01-16T04:20:00
    lon         (trajectory) float32 ...
    lat         (trajectory) float32 ...

* 注意若使用`.concat`合并的将会是追加，append的操作

In [42]:
xar.concat([xr_temp,xr_temp_y],dim='trajectory')

NameError: name 'xr_temp' is not defined

以下出现错误：
 MergeError: conflicting values for variable 'time' on objects to be combined. You can skip this check by specifying compat='override'.  
 出现此错误经检查为 `time` 不是同一个time导致的  
 #### 最终使用merage的方式将两个维度进行合并

In [43]:
xr_merge=xar.merge([xr_temp_x])

In [44]:
xr_merge

<xarray.Dataset>
Dimensions:     (trajectory: 2958)
Coordinates:
  * trajectory  (trajectory) int32 1 2 3 4 5 6 ... 2953 2954 2955 2956 2957 2958
    time        datetime64[ns] 2018-01-16T04:20:00
    lon         (trajectory) float32 ...
    lat         (trajectory) float32 ...
Data variables:
    x_wind      (trajectory) float32 9.96921e+36 -2.8568854 ... 9.96921e+36

In [45]:
xr_merge=xar.merge([xr_temp_x,xr_temp_y])
xr_merge

<xarray.Dataset>
Dimensions:     (trajectory: 2958)
Coordinates:
  * trajectory  (trajectory) int32 1 2 3 4 5 6 ... 2953 2954 2955 2956 2957 2958
    time        datetime64[ns] 2018-01-16T04:20:00
    lon         (trajectory) float32 9.96921e+36 125.85207 ... 9.96921e+36
    lat         (trajectory) float32 9.96921e+36 28.504345 ... 9.96921e+36
Data variables:
    x_wind      (trajectory) float32 9.96921e+36 -2.8568854 ... 9.96921e+36
    y_wind      (trajectory) float32 9.96921e+36 10.472756 ... 9.96921e+36

In [47]:
# 剔除大于阈值的，并且为nan的
df_finial=xr_merge.where(xr_merge<2000).to_dataframe().dropna(how='any')
df_finial

x_wind     y_wind                time         lon        lat
trajectory                                                                
2          -2.856885  10.472756 2018-01-16 04:20:00  125.852074  28.504345
4          -3.510324   8.823525 2018-01-16 04:20:00  125.837547  28.535669
5          -3.477507   8.365860 2018-01-16 04:20:00  125.746597  28.534342
9          -3.867782   9.866551 2018-01-16 04:20:00  125.779015  28.548067
10         -4.172445   8.833652 2018-01-16 04:20:00  125.941368  28.543129
...              ...        ...                 ...         ...        ...
1250       -2.741773  11.013288 2018-01-16 04:20:00  125.921379  28.369308
1251       -4.936096   8.763128 2018-01-16 04:20:00  125.921432  28.369282
1252       -1.048339  10.038569 2018-01-16 04:20:00  125.921486  28.369307
1253       -3.387036   9.787621 2018-01-16 04:20:00  125.921524  28.369263
1254       -3.869750  10.290464 2018-01-16 04:20:00  125.921501  28.369329

[901 rows x 5 columns]

## 3 将df写入数据库

In [48]:
type(df_finial)

pandas.core.frame.DataFrame

注意此处以后不再使用`.ix`进行索引，使用`iloc`进行索引

In [49]:
df_finial.ix[2]

D:\01Setup\ANACONDA\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


x_wind               -2.85689
y_wind                10.4728
time      2018-01-16 04:20:00
lon                   125.852
lat                   28.5043
Name: 2, dtype: object

In [52]:
len(df_finial)

901

In [50]:
df_finial.iloc[0]

x_wind               -2.85689
y_wind                10.4728
time      2018-01-16 04:20:00
lon                   125.852
lat                   28.5043
Name: 2, dtype: object

In [51]:
df_finial.iloc[0].get('x_wind')

-2.8568854